### DISCUSSION TITLE
Bandwidth limitation overview and troubleshooting

### DISCUSSION THREAD

**Discussion by rechia-venko** [magma/magma#15292] [Q&A] [answered]
Hello fellow magma users and contributors!

We have the objective of setting up different data plans for different subscribers depending on the amount of data the subscriber has used up so far. For instance:

1. Subscriber has 5 Mbps symmetric data plan up until they consume 1 GB of data
2. As soon as the Sub hits 1 GB of data, we'd like to swap the data plan to 1 Mbps max bandwidth and **immediately** limit the bandwidth available to this subscriber.

However, during our tests the `sub_profile` change is not working for us. The bandwidth is not limited at all according to the data plan. In fact, the only bandwidth limitation that works for us is the APN limit. Here are some of the things we have configured:

- APN max bandwidth set to 10 Mbps
```c
{ apn_ambr_10mbps: {
  'apn_configuration': {
    'ambr': {
      'max_bandwidth_dl': 10000000, 'max_bandwidth_ul': 10000000
            }, 
      'qos_profile': {'class_id': 9, 'preemption_capability': False, 'preemption_vulnerability': False, 'priority_level': 15
            }
        }, 'apn_name': 'apn_ambr_10mbps'
    }
}
```
![image](https://github.com/magma/magma/assets/135657713/9c594c0f-7415-4c55-8dbb-66cc18f89579)

- Base name, policy rules, and QoS profile
![image](https://github.com/magma/magma/assets/135657713/daaeada8-583a-4af7-823f-f69b233431bf)
![image](https://github.com/magma/magma/assets/135657713/1b910ca2-a179-45df-b0be-bc1e1d9ef52c)
![image](https://github.com/magma/magma/assets/135657713/48ca8e4e-8db8-4716-8aa4-187e996ad307)
 

- data plans:
![image](https://github.com/magma/magma/assets/135657713/3d293d4a-8c20-4ba2-bbaa-085e79aa7b63)

And the subscriber under test is this one:

![image](https://github.com/magma/magma/assets/135657713/9f15ca2a-39ca-45ba-85fb-814f14e73d33)

**The only bandwidth limitation we can enforce for this sub is the APN limit at 10Mbps or 20 Mbps depending on the APN selected.**

How can we troubleshoot this? We have many questions:
- How can I check OVS and tc rules for each bandwidth limiting entity?
  - QoS Profile
  - APN ambr
  - policy rule
  - Data plan
- What are base names used for. We can see that we can associate many subscribers to a single base name. What is the purpose of this? Is it to enforce a selected QoS profile for all these subscribers? It doesn't seem to work for us either.
- Where can I find more info about all these bandwidth limiting entities and how to configure them? If I configure limitations on several places, how are these limitations prioritized?
- How can we make the data plan swap immediate? Do we have to restart AGW services to make sure that a new data plan takes place immediately?

**Comment by rechia-venko** [comment]
I've browsed and read this other discussion https://github.com/magma/magma/discussions/5816. It is related, but not exactly the same set of questions.

Here we need to better understand how to configure and troubleshoot the many ways of limiting bandwidth, and also better understand how these limitation mechanisms interact with each other. We also need to apply modifications to the bandwidth immediately (we intend to use the orc8r api to do this)

**Comment by javieraubert** [comment]
Hello, just to be sure I'm understanding right, no limitation is being applied at all, or just when the data plan amount is exceeded?

**Reply by rechia-venko** [reply]
No limitation is being applied at all!

**Comment by dishnetafrica** [comment]
To apply a higher bandwidth limit you need to configure that the user has to reconnect after consuming 1 GB.to achieve this you need external billing platform. you may contact mycloudbss who has platform to apply this changes . https://mycloudbss.com/ +91 8511976588 (whats app)

**Reply by rechia-venko** [reply]
Our end goal is to write our own billing app, **but the data plan limitation is not working in the first place**. We need help to troubleshoot why the data plan bandwidth limitation does not work.

Later on we'll also need to understand how to apply the data limitation in real-time without having to disconnect the user. But that is a problem for the future.

**Reply by Darlanewe** [reply]
are you brazillian by any chance?

**Reply by Darlanewe** [reply]
legal ver compatriotas por aqui, 
espero que possamos colaborar pra difundir o magma no br

**Reply by lucasgonze** [reply]
I believe you have found a bug or bugs and this should become a work item to fix.

**Comment by javieraubert** [comment]
Hello @rechia-venko did you make it work?

**Comment by lucasgonze** [comment]
Todo: write a bug. Prioritize in 2.10.

In the meantime, it's not a question.

**Reply by lucasgonze** [reply, marked as answer]
Created issue https://github.com/magma/magma/issues/15362
